UploadDaemon/SymbolAnalysis/SymbolCollectionResolver.cs

```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Collections.Generic;
+using UploadDaemon.Configuration;
+
+namespace UploadDaemon.SymbolAnalysis
+{
+    /// <summary>
+    /// This resolver loads symbol collections (PDB files) into <see cref="SymbolCollection"/>s.
+    /// It uses a cache to avoid unnecessary I/O from reading the same PDB files repeatedly.
+    /// </summary>
+    public class SymbolCollectionResolver
+    {
+        private readonly IDictionary<Tuple<string, GlobPatternList>, SymbolCollection> symbolCollectionsCache =
+            new Dictionary<Tuple<string, GlobPatternList>, SymbolCollection>();
+
+        public SymbolCollection ResolveFrom(string symbolDirectory, GlobPatternList assemblyPatterns)
+        {
+            Tuple<string, GlobPatternList> key = GetCacheKey(symbolDirectory, assemblyPatterns);
+            if (!symbolCollectionsCache.ContainsKey(key))
+            {
+                symbolCollectionsCache[key] = SymbolCollection.CreateFromPdbFiles(symbolDirectory, assemblyPatterns);
+            }
+            return symbolCollectionsCache[key];
+        }
+
+        private static Tuple<string, GlobPatternList> GetCacheKey(string symbolDirectory, GlobPatternList assemblyPatterns)
+        {
+            return (symbolDirectory, assemblyPatterns).ToTuple();
```

Comment:
calling ToTuple on a tuple seems kind of redundant. Can't you just change the return type to `(string, GlobPatternList)` and use the same type in the map? Or is there a technical reason why that won't work?

Reply:
The problem is that `(symbolCOllection, assemblyPatterns)` is, in fact, not a `Tuple`. Using `(string, GlobalPatternList)` works, though. I wasn't aware that this is a valid type (declaration).